worlds/yachtdice/Rules.py

```diff
@@ -0,0 +1,223 @@
+import math
+from collections import defaultdict
+
+#List of categories, and the name of the logic class associated with it
+category_mappings = {
+    "Category Ones": "Ones",
+    "Category Twos": "Twos",
+    "Category Threes": "Threes",
+    "Category Fours": "Fours",
+    "Category Fives": "Fives",
+    "Category Sixes": "Sixes",
+    "Category Choice": "Choice",
+    "Category Inverse Choice": "Choice",
+    "Category Pair": "Pair",
+    "Category Three of a Kind": "ThreeOfAKind",
+    "Category Four of a Kind": "FourOfAKind",
+    "Category Tiny Straight": "TinyStraight",
+    "Category Small Straight": "SmallStraight",
+    "Category Large Straight": "LargeStraight",
+    "Category Full House": "FullHouse",
+    "Category Yacht": "Yacht",
+  
+    "Category Distincts": "Distincts",
+    "Category Two times Ones": "Twos", #same weights as twos category
+    "Category Half of Sixes": "Threes", #same weights as threes category
+    "Category Twos and Threes": "TwosAndThrees",
+    "Category Sum of Odds": "SumOfOdds",
+    "Category Sum of Evens": "SumOfEvens",
+    "Category Double Threes and Fours": "DoubleThreesAndFours",
+    "Category Quadruple Ones and Twos": "QuadrupleOnesAndTwos",
+    "Category Micro Straight": "MicroStraight",
+    "Category Three Odds": "ThreeOdds",
+    "Category 1-2-1 Consecutive": "OneTwoOneConsecutive",
+    "Category Three Distinct Dice": "ThreeDistinctDice",
+    "Category Two Pair": "TwoPair",
+    "Category 2-1-2 Consecutive": "TwoOneTwoConsecutive",
+    "Category Five Distinct Dice": "FiveDistinctDice",
+    "Category 4&5 Full House": "FourAndFiveFullHouse"
+}
+
+#This class adds logic to the apworld.
+#In short, we ran a simulation for every possible combination of dice and rolls you can have, per category.
+#This simulation has a good strategy for locking dice.
+#This gives rise to an approximate discrete distribution per category.
+#We calculate the distribution of the total score.
+#We then pick a correct percentile to reflect the correct score that should be in logic.
+#The score is logic is *much* lower than the actual maximum reachable score.
+
+
+class Category:
+    def __init__(self, name, quantity = 1):
+        self.name = name
+        self.quantity = quantity #how many times you have the category
+
+    #return mean score of a category
+    def mean_score(self, num_dice, num_rolls):
+        if num_dice == 0 or num_rolls == 0:
+            return 0
+        mean_score = 0
+        for key in yacht_weights[self.name, min(8,num_dice), min(8,num_rolls)]:
+            mean_score += key*yacht_weights[self.name, min(8,num_dice), min(8,num_rolls)][key]/100000
+        return mean_score * self.quantity
+
+
+def extract_progression(state, player, options):
+    #method to obtain a list of what items the player has.
+    #this includes categories, dice, rolls and score multiplier etc.
+    
+    if player == "state_is_a_list": #the state variable is just a list with the names of the items
```

Comment:
So, as far as I understand it, every check in the game uses the same "custom state", which is what you calculate in `extract_progression`. I'll keep referring to return value of `extract_progression` as the "custom state" from here on out.

So, in any given location's access rule, the maximum achievable score is calculated and compared.
The maximum achievable score calculation is cached by the custom state. Very good.
However, and correct me if I'm wrong, but the custom state is not retained or cached in any way.

In my opinion, you should make the custom state (or maximum score, for that matter) persistent, and **override state.collect() and state.remove()**, and do the calculation of the custom state every time they get called with any (relevant) item - I.e. "on state change".

If you want to be really fancy, instead of recalcing the state every time `collect` or `remove` gets called, your overriden `state.collect()` and `state.remove()` could instead just set a custom_state_is_stale variable to `True`, and then check for that variable at the start of `extract_progression` to see if the custom state needs to be recalculated. 

I hope this makes sense.

Reply:
Btw, I would also recommend changing the "custom state" to a dataclass or NamedTuple, that would make it a bit easier to follow in my opinion. The dataclass could have an overriden str() method that then replaces your string representation conversion

Reply:
In fact, yeah, I think I'm convincing myself that this is the correct approach:

1. Make `maximum_achievable_score` a persistent state property, as it's the only thing that actually "matters" in the end
2. Make a second persistent state property called `achievable_score_is_stale`. If this is True, the maximum score needs to be recalculated.
3. Override collect & remove to set `achievable_score_is_stale` to True.

Reply:
Full disclosure, I've never done the collect/remove overriding myself, so I might be missing some nuance.

Reply:
Thank you for the suggestions, I'll get working on it later today!
Yes you seem to understand how the code works!

I want to double check if I understand your suggestions.

First, here's some extra information.
I did a test run with Yacht Dice, SM64 and Celeste64, and here's the number of calls to the Yacht Dice logic:
- 29341 calls to `dice_simulation_strings`
- Out of all these calls, in 27386 of them, the state (of progression items, excl extra points) was exactly the same as the previous time the function was called
- In total, `dice_simulation_strings` was called with 296 unique states (of progression items, excl extra points)

I agree with your latest approach. One thing I'm planning to add to it, is that I only change `achievable_score_is_stale` to True, if a progressive item is added. 

I will also not change it to True when an Extra Point is added or removed, I will keep these outside of "maximum_achievable_score", and will make the logic be something like 
```
score_to_beat = 100 #example
extra_points_in_logic = state.count("1 Point", player)
extra_points_in_logic += state.count("10 Points", player) * 10
extra_points_in_logic += state.count("100 Points", player) * 100
if achievable_score_is_stale:
  update maximum achievable score
return state.has("maximum_achievable_score", player, score_to_beat - extra_points_in_logic)
```

Does this make sense to you?

Reply:
Yeah that seems fine :) state.count is quite fast.

If you ever have more denominations of point values though, it might be worth handling that through collect and remove as well, in a similar way to how DLCQuest works.

Reply:
I've changed the extract_progression in Rules so that it doesn't do everything twice.
And as you know your idea works :)